### PR TITLE
Generate ELF shared libraries and allow multi-abi libs in APKs and App bundles

### PR DESCRIPTION
**Pull request by blasten**
## Description

Include ELF shared library instead of AOT snapshots in APKs and make the flag `target-platform` multi-option for `build apk` and `build appbundle`.

After this change, an APK built for release or profile will have the following structure:
```diff
    2460  Defl:N      889  64% 1980-00-00 00:00 f40cf37a  AndroidManifest.xml
     765  Defl:N      595  22% 1980-00-00 00:00 dce0efa9  META-INF/CERT.RSA
    1522  Defl:N      722  53% 1980-00-00 00:00 39881335  META-INF/CERT.SF
    1479  Defl:N      695  53% 1980-00-00 00:00 ce61d107  META-INF/MANIFEST.MF
     109  Defl:N       53  51% 1980-00-00 00:00 e83e6b87  assets/flutter_assets/AssetManifest.json
     208  Defl:N      113  46% 1980-00-00 00:00 92d58e5f  assets/flutter_assets/FontManifest.json
  644419  Defl:N    58421  91% 1980-00-00 00:00 548a9a0b  assets/flutter_assets/LICENSE
  134640  Defl:N    63011  53% 1980-00-00 00:00 c7c94b78  assets/flutter_assets/fonts/MaterialIcons-Regular.ttf
   97680  Defl:N    54923  44% 1980-00-00 00:00 273d62de  assets/flutter_assets/packages/cupertino_icons/assets/CupertinoIcons.ttf
- 1821616  Defl:N   773152  58% 1980-00-00 00:00 10467cf1  assets/isolate_snapshot_data
- 3060736  Defl:N   832709  73% 1980-00-00 00:00 0406652a  assets/isolate_snapshot_instr
-   23592  Defl:N    11409  52% 1980-00-00 00:00 0fa678e0  assets/vm_snapshot_data
-   12704  Defl:N     2807  78% 1980-00-00 00:00 b5e3f0f0  assets/vm_snapshot_instr
  258100  Defl:N   107181  59% 1980-00-00 00:00 5bc8eac3  classes.dex
+ 5632000  Defl:N  1850034  67% 1980-00-00 00:00 6415dcf7  lib/armeabi-v7a/libapp.so
 6032316  Defl:N  3293989  45% 1980-00-00 00:00 c82f4361  lib/armeabi-v7a/libflutter.so
     344  Defl:N      186  46% 1980-00-00 00:00 55202761  res/drawable/launch_background.xml
     507  Stored      507   0% 1980-00-00 00:00 dab4e8f1  res/mipmap-hdpi-v4/ic_launcher.png
     405  Stored      405   0% 1980-00-00 00:00 ad6e0a73  res/mipmap-mdpi-v4/ic_launcher.png
     684  Stored      684   0% 1980-00-00 00:00 77d229a5  res/mipmap-xhdpi-v4/ic_launcher.png
     994  Stored      994   0% 1980-00-00 00:00 547189ae  res/mipmap-xxhdpi-v4/ic_launcher.png
    1406  Stored     1406   0% 1980-00-00 00:00 63bc7c87  res/mipmap-xxxhdpi-v4/ic_launcher.png
```

### Fat APK

```sh
$ flutter build apk

# Effectively the same as
$ flutter build apk --target-platform android-arm,android-arm64

# Also possible in the near future
$ flutter build apk --target-platform android-arm,android-arm64,android_x64
```

The tool displays a message:

```
You are building a fat APK that includes binaries for android-arm, android-arm64.
If you are deploying the app to the Play Store, it's recommended to use app bundles or split the APK to reduce the APK size.
    To generate an app bundle, run:
        flutter build appbundle --target-platform android-arm,android-arm64
        Learn more on: https://developer.android.com/guide/app-bundle
    To split the APKs per ABI, run:
        flutter build apk --target-platform android-arm,android-arm64 --split-per-abi
        Learn more on:  https://developer.android.com/studio/build/configure-apk-splits#configure-abi-split
```

Resulting APK:

```diff
    2512  Defl:N      901  64% 00-00-1980 00:00 c7f247d8  AndroidManifest.xml
     765  Defl:N      597  22% 00-00-1980 00:00 b105c0a9  META-INF/CERT.RSA
    1678  Defl:N      777  54% 00-00-1980 00:00 45833d1e  META-INF/CERT.SF
    1635  Defl:N      747  54% 00-00-1980 00:00 a3af0ca9  META-INF/MANIFEST.MF
     109  Defl:N       53  51% 00-00-1980 00:00 e83e6b87  assets/flutter_assets/AssetManifest.json
     208  Defl:N      113  46% 00-00-1980 00:00 92d58e5f  assets/flutter_assets/FontManifest.json
  644419  Defl:N    58421  91% 00-00-1980 00:00 548a9a0b  assets/flutter_assets/LICENSE
  134640  Defl:N    63011  53% 00-00-1980 00:00 c7c94b78  assets/flutter_assets/fonts/MaterialIcons-Regular.ttf
   97680  Defl:N    54923  44% 00-00-1980 00:00 273d62de  assets/flutter_assets/packages/cupertino_icons/assets/CupertinoIcons.ttf
  257176  Defl:N   107353  58% 00-00-1980 00:00 8e495fa1  classes.dex
 6311936  Defl:N  1801944  72% 00-00-1980 00:00 ed1058f5  lib/arm64-v8a/libapp.so
 8473344  Defl:N  3657267  57% 00-00-1980 00:00 540aacd2  lib/arm64-v8a/libflutter.so
 5632000  Defl:N  1850033  67% 00-00-1980 00:00 8abc54a5  lib/armeabi-v7a/libapp.so
 6032316  Defl:N  3293989  45% 00-00-1980 00:00 c82f4361  lib/armeabi-v7a/libflutter.so
     344  Defl:N      186  46% 00-00-1980 00:00 55202761  res/drawable/launch_background.xml
     507  Stored      507   0% 00-00-1980 00:00 dab4e8f1  res/mipmap-hdpi-v4/ic_launcher.png
     405  Stored      405   0% 00-00-1980 00:00 ad6e0a73  res/mipmap-mdpi-v4/ic_launcher.png
     684  Stored      684   0% 00-00-1980 00:00 77d229a5  res/mipmap-xhdpi-v4/ic_launcher.png
     994  Stored      994   0% 00-00-1980 00:00 547189ae  res/mipmap-xxhdpi-v4/ic_launcher.png
    1406  Stored     1406   0% 00-00-1980 00:00 63bc7c87  res/mipmap-xxxhdpi-v4/ic_launcher.png
    1560  Stored     1560   0% 00-00-1980 00:00 71db73fd  resources.arsc
```

### ABI split per APK

```sh
$ flutter build apk --split-per-abi
```
This will generate two APKs with the appropriate version for the the Play Store:

* Built __build/app/outputs/apk/release/app-armeabi-v7a-release.apk__.
* Built __build/app/outputs/apk/release/app-arm64-v8a-release.apk__.
 

## Related Issues

https://github.com/flutter/flutter/issues/18494
https://github.com/flutter/flutter/issues/31922

## Tests

* Unit tests to the AOT snapshotter.
* Unpack the APK and app bundle and assert that the expected libraries are contained.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by jason-simmons — APPROVED** [review]

**Comment by goderbauer** [comment]
Wondering what this change does to APK size. From the diff in the description it looks like it adds ~220KiB compressed?

**Comment by jason-simmons** [comment]
I tried building a default template app with `flutter build apk --release` and `flutter build apk --release --build-shared-library`

The regular APK was 5.23MB and the `build-shared-library` APK was 5.46MB.

The uncompressed `app.so` library was 5.96MB.  Versus 4.97MB for the total of `isolate_snapshot_data`, `isolate_snapshot_instr`, `vm_snapshot_data`, and `vm_snapshot_instr` in the regular APK.

Note that we had to disable the debug symbol stripping that Android usually does to `.so` libraries packaged into an APK.  See https://github.com/dart-lang/sdk/issues/37146

@rmacnak-google

**Comment by goderbauer** [comment]
Is the size increase likely due to the debug symbols?

**Comment by blasten** [comment]
FYI: the config to avoid stripping debug symbols does not work in Flutter modules. It seems that the root cause is that the Gradle plugin is inserted as an implementation dependency:
```gradle
dependencies {
    implementation project(':flutter')
}
```
In which, any config set in the `android { }` block by the plugin is completely ignored. The app's `.android/app/build.gradle` needs to set `doNotStrip`.

We may need to land a fix for dart-lang/sdk#37146 before this one goes in.

cc @rmacnak-google @jason-simmons @tvolkert

**Comment by blasten** [comment]
The CI seems to be running into: https://github.com/dart-lang/sdk/blob/a6bbc58d09569ed84f437d6ff55c01166745b3d4/runtime/vm/dart_api_impl.cc#L6075. The target is still `arm`, but the command is run from `Windows`. Is that expected?

**Comment by blasten** [comment]
Pending AIs: 
* [ ] The issue about Android stripping symbols was fixed in https://github.com/dart-lang/sdk/commit/da0de06643a4cac5f190c353180491e8db85c6e1 (Confirmation pending, AI: remove `doNotStrip "*/${abi}/libapp.so"` from Gradle).
* [ ] The failure on Windows will be fixed by https://github.com/flutter/engine/pull/9190/files
* [ ] Add integration test that installs an APK in release mode from `build apk` for  app and add to app.
* [ ] Add integration test that installs an APK in release mode from `build appbundle`. Requires to update the image in devicelab to include `bundletool` for app and add to app.

**Comment by tvolkert** [comment]
> The CI seems to be running into: https://github.com/dart-lang/sdk/blob/a6bbc58d09569ed84f437d6ff55c01166745b3d4/runtime/vm/dart_api_impl.cc#L6075. The target is still arm, but the command is run from Windows. Is that expected?

@rmacnak-google

**Comment by blasten** [comment]
@tvolkert  I think https://github.com/flutter/engine/pull/9190/files will fix it.

**Comment by jason-simmons** [comment]
See also https://github.com/flutter/buildroot/pull/258 and https://dart-review.googlesource.com/c/sdk/+/104950

File: dev/devicelab/bin/tasks/gradle_plugin_test.dart
Comment (tvolkert) [inline dev/devicelab/bin/tasks/gradle_plugin_test.dart]: How come x86 and x86_64 are in here too if we specified `'-Ptarget-platform=android-arm'`?

Comment (tvolkert) [reply dev/devicelab/bin/tasks/gradle_plugin_test.dart]: Maybe because it's debug mode, and we proactively include them for simulators?

File: dev/devicelab/bin/tasks/gradle_plugin_test.dart
Comment (tvolkert) [inline dev/devicelab/bin/tasks/gradle_plugin_test.dart]:
```dart
return await unzip.stdout
    .transform(utf8.decoder)
    .transform(const LineSplitter())
    .map((String line) => line.split(' ').last)
    .toList();
```

File: dev/devicelab/bin/tasks/gradle_plugin_test.dart
Comment (tvolkert) [inline dev/devicelab/bin/tasks/gradle_plugin_test.dart]: No need for `async`

File: dev/devicelab/bin/tasks/gradle_plugin_test.dart
Comment (blasten) [reply dev/devicelab/bin/tasks/gradle_plugin_test.dart]: Done.

File: dev/devicelab/bin/tasks/gradle_plugin_test.dart
Comment (blasten) [reply dev/devicelab/bin/tasks/gradle_plugin_test.dart]: Yup. I added a comment to the test :)

File: dev/devicelab/bin/tasks/gradle_plugin_test.dart
Comment (blasten) [reply dev/devicelab/bin/tasks/gradle_plugin_test.dart]: Done.

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (tvolkert) [inline packages/flutter_tools/lib/src/android/gradle.dart]: This might choose an APK that doesn't match the device's arch, no?  Seems like we probably need to refactor our logic that currently looks in `project.apkDirectory.childFile('app.apk')` to handle multiple APKs and choose the right one when installing, running, etc.

(beyond the scope of this change though - can be a TODO for now)

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (tvolkert) [inline packages/flutter_tools/lib/src/android/gradle.dart]: Nit: `return const <File>[];`

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (tvolkert) [inline packages/flutter_tools/lib/src/android/gradle.dart]: `project.apkDirectory.childFile(apkFileName)`

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (tvolkert) [inline packages/flutter_tools/lib/src/android/gradle.dart]: `project.apkDirectory.childDirectory(modeName).childFile(apkFileName)`

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (tvolkert) [inline packages/flutter_tools/lib/src/android/gradle.dart]: With this many subdirectories, this way is actually shorter, but the other way allows chaining... I think the chaining is easier to read, but your call.

```dart
project.apkDirectory
    .childDirectory(buildInfo.flavor)
    .childDirectory(modeName)
    .childFile(apkFileName)
```

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (tvolkert) [inline packages/flutter_tools/lib/src/android/gradle.dart]: Nit: `const <String>[]`

File: packages/flutter_tools/lib/src/build_info.dart
Comment (tvolkert) [inline packages/flutter_tools/lib/src/build_info.dart]: `///`

Comment (tvolkert) [reply packages/flutter_tools/lib/src/build_info.dart]: Can you also add more here about what this means. possibly something like:

```dart
/// Whether to split the build artifacts per ABI.
///
/// When this is false, multiple ABIs will be contained within one primary
/// build artifact. When this is true, multiple build artifacts (one per ABI)
/// will be produced.
```

File: packages/flutter_tools/lib/src/build_info.dart
Comment (tvolkert) [inline packages/flutter_tools/lib/src/build_info.dart]: Can you document valid values?  e.g. is this guaranteed to be non-empty?  Or can it be empty, but that signals invalid build info?

File: packages/flutter_tools/lib/src/runner/flutter_command.dart
Comment (tvolkert) [inline packages/flutter_tools/lib/src/runner/flutter_command.dart]: `build_aot.dart` and `build_bundle.dart` both still specify a single `--target-platform` argument, which if it hits this code path will cause a runtime error trying to call `map()` on `String`.

File: packages/flutter_tools/lib/src/runner/flutter_command.dart
Comment (tvolkert) [inline packages/flutter_tools/lib/src/runner/flutter_command.dart]: Nit: `const <TargetPlatform>[]`

File: dev/devicelab/bin/tasks/gradle_plugin_test.dart
Comment (blasten) [reply dev/devicelab/bin/tasks/gradle_plugin_test.dart]: Done.

File: packages/flutter_tools/gradle/flutter.gradle
Comment (blasten) [reply packages/flutter_tools/gradle/flutter.gradle]: I think so. One thing that is worth doing later is to remove platforms and only have archs. It simplifies the terminology and unnecessary mappings.

File: packages/flutter_tools/gradle/flutter.gradle
Comment (blasten) [reply packages/flutter_tools/gradle/flutter.gradle]: I added a comment. The default behavior is to build a fat APK.

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (blasten) [reply packages/flutter_tools/lib/src/android/gradle.dart]: Added the TODO.

File: packages/flutter_tools/lib/src/android/gradle.dart
Comment (blasten) [reply packages/flutter_tools/lib/src/android/gradle.dart]: +1

File: packages/flutter_tools/lib/src/build_info.dart
Comment (blasten) [reply packages/flutter_tools/lib/src/build_info.dart]: So, I ended up refactoring this code a little bit. Now, there's `AndroidBuildInfo`which contains a `BuildInfo`, `targetArchs` and `splitPerAbi`.   I defined `AndroidArch`, which mirrors `IosArch`.

Long term goal is to remove the concept of `android-arm`, `android-arm64`..., and instead use the abi values directly: https://developer.android.com/ndk/guides/abis

This makes the code self-explanatory and avoids combinations that aren't used in other target platforms. Indeed, the current `targetPlatform` in `BuildInfo` only makes sense in Android to pass arm, arm64 to gradle.

File: packages/flutter_tools/lib/src/runner/flutter_command.dart
Comment (blasten) [reply packages/flutter_tools/lib/src/runner/flutter_command.dart]: This isn't an issue anymore since I refactored the code, and `BuildInfo` doesn't have target platform anymore.

File: packages/flutter_tools/gradle/flutter.gradle
Comment (mklim) [inline packages/flutter_tools/gradle/flutter.gradle]: This is another very subjective nit, feel free to ignore.

If you flip the structure of this into an early return you end up with most of the logic being outside of a block instead of inside of one, so you don't have to keep as much nesting in mind while you're reading the function and trying to figure out the control flow. Subjective but it helps readability a bit in my opinion. This applies in a couple places.

```groovy
if (!project.hasProperty('target-platform')) {
  return DEFAULT_PLATFORMS
}
Set<String> platforms = []; 
project.property('target-platform').split(',').each { // etc
```

https://testing.googleblog.com/2017/06/code-health-reduce-nesting-reduce.html

Also a really minor optional nit, but I think mapping here would be a little cleaner than doing a for-each loop and then appending to a list. The groovy version of map is [`collect()`](http://docs.groovy-lang.org/2.4.3/html/groovy-jdk/java/util/Collection.html#collect(groovy.lang.Closure)).

File: dev/devicelab/bin/tasks/gradle_plugin_test.dart
Comment (blasten) [reply dev/devicelab/bin/tasks/gradle_plugin_test.dart]: This will return a `Stream`, which isn't a subtype of `Iterable`.

File: packages/flutter_tools/gradle/flutter.gradle
Comment (mklim) [inline packages/flutter_tools/gradle/flutter.gradle]: This is going to prevent people from transparently controlling the version code of their APKs. I don't think it's a breaking change exactly but it may impact developers who are already relying on a specific versioning scheme like https://developer.android.com/google/play/publishing/multiple-apks.html#using-a-version-code-scheme, or who are relying on their builds matching a particular version code that's been uploaded to the Play Developer Console. Can there be some documentation about this where the flag is defined so people aren't surprised?

File: dev/devicelab/bin/tasks/gradle_plugin_test.dart
Comment (mklim) [reply dev/devicelab/bin/tasks/gradle_plugin_test.dart]: I think this fix got missed somewhere in git history.

**Review by mklim — APPROVED** [review]
LGTM pending LGTM from other reviewers as well. :)

File: packages/flutter_tools/gradle/flutter.gradle
Comment (tvolkert) [inline packages/flutter_tools/gradle/flutter.gradle]: I think this is two comments as one (we probably only need to keep the "Resets the list of ABIs ..." line.

**Review by tvolkert — APPROVED** [review]
Other than a few nits, LGTM

File: packages/flutter_tools/gradle/flutter.gradle
Comment (blasten) [reply packages/flutter_tools/gradle/flutter.gradle]: Yep. It's documented in `flutter help build apk`. It's an optional flag.

File: packages/flutter_tools/gradle/flutter.gradle
Comment (blasten) [reply packages/flutter_tools/gradle/flutter.gradle]: +1

File: dev/devicelab/bin/tasks/gradle_plugin_test.dart
Comment (blasten) [reply dev/devicelab/bin/tasks/gradle_plugin_test.dart]: I see your point. Also, relevant: https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html.

**Comment by blasten** [comment]
@tvolkert 

> How do we know it's a profile or release build at this point?

There's a check before that.  I have updated both error messages, so it's more clear.

**Review by jonahwilliams — APPROVED** [review]
LGTM

File: packages/flutter_tools/gradle/flutter.gradle
Comment (tvolkert) [inline packages/flutter_tools/gradle/flutter.gradle]: Can this ever be null anymore? (is the parameter truly optional?)

**Review by tvolkert — APPROVED** [review]
LGTM

File: packages/flutter_tools/gradle/flutter.gradle
Comment (blasten) [reply packages/flutter_tools/gradle/flutter.gradle]: It extracts `libflutter.so` as well, then it adds the `libflutter.so` for the rest of the ABIs.
Alternatively, we could extract the `.class` only before the `afterEvaluate` block (otherwise Gradle complains that we are referencing classes that aren't defined).

I filed an issue to track this in https://github.com/flutter/flutter/issues/33059 and https://github.com/flutter/flutter/issues/33064

The current engine artifacts for Android are:
```
android-arm/
android-arm-profile/
android-arm-release/
android-arm64/
android-arm64-profile/
android-arm64-release/
android-x64/
android-x86/
```

`x64` and `x86` are for debug only, and their corresponding `libflutter.so` is always added even for `arm` and `arm64`.

File: packages/flutter_tools/gradle/flutter.gradle
Comment (blasten) [reply packages/flutter_tools/gradle/flutter.gradle]: 👍

**Comment by jonahwilliams** [comment]
This change has doubled the measured size in the basic_material_app_android__compile benchmark.

**Comment by blasten** [comment]
@jonahwilliams that's intended. Looks like the goal/baseline is defined in a db?

**Pull request merged**
